### DISCUSSION TITLE
[Feature Request] Creating digital stamps on documents

### DISCUSSION THREAD

**Discussion by tinyoverflow** [paperless-ngx/paperless-ngx#3744] [Feature Requests]
### Description

Hello,

I'd like to suggest a stamping feature. Many proprietary DMS have that and it allows the user to stamp a document digitally, just like with a real stamp. This stamp would be embedded onto the document.

**Use case**
It is typically used for stamps like "PAID [date], [bank account]" or "DELIVERED [date]".
I personally stamp all my incoming documents with a so called Eingangsstempel (german), which stamps the date of when I received the letter onto the document.

Then, for example, when I have to pay something, I put a stamp on it again with "PAID [date of payment]", so I always have it right on the document in case I need that info.

This feature could also be used for signing a document digitally (like not real certificate signing, but signing it with your name).

**Current alternative and it's drawbacks**
Of course, all those infos could be put into the note field in Paperless, but these informations are then lost when the document gets printed and handed over or sent to someone else. These digital stamps would be printed onto the document, where the user placed them.

**Notable**
In my opinion the stamps must not be transferred over to the original document PDF, but only to the edited one by Paperless.

What do you think about this?

**Comment by alabiana** [comment]
This would be very cool

**Comment by DO10-01** [comment]
A stamp function in the PDF file, like in Acrobat, would be an important and very helpful function. The absence is the reason why I don't scan directly into the Consume folder, but into an input folder on the hard drive. There, the documents receive an entry stamp and are only then moved from the local hard drive to the Consume folder. That still doesn't solve the problem with other stamps like "paid on" et cetera, but at least there's an entry stamp on it. For many months, the incoming paper has disappeared directly into the archive after being scanned.
[automatically translated]

**Comment by madduck** [comment]
Instead of modifying the PDF, we really ought to be attaching cryptographic claims. I.e. the claim "I paid this on xx" signed with my key.

**Reply by tinyoverflow** [reply]
I think that this does not solve the problem. The cryptographic key does nothing when we print the document. This would have the same downside as just typing the stamp-note into the notes field. 

This is the main reason for stamps: Having it on the document. The original document wouldn‘t be manipulated anyway, as it is stored separately in Paperless.

**Reply by madduck** [reply]
I understand where you're coming from, but I still think that we *must not* modify the PDF. After all, it's PDF/A-2 for a reason, and there are legal requirements in place for that in parts of this world, e.g. Germany.

The benefit of attaching cryptographic metadata is twofold IMHO:

1. You can verify it, and link a claim back to a person, which you cannot if all you do is stamp a document;
2. The metadata *can* be used to *create an export* of the document, with the information in whatever format is desirable at the time and for the purpose of the export, i.e. a stamp printed on the PDF, or maybe an extra page, or maybe you just need that extra JSON field in the result of an API call. If you modify the document rather than its metadata, you presumed this choice, and it'll be hard(er) later to (a) restore the original, and (b) extract the information in another format.

**Reply by DO10-01** [reply]
It would certainly be laborious, but could it perhaps be done like Adobe Lightroom? Non-destructive. All changes are visibly made to the display, the file on disk not changed and the changes saved to the database. They could then be applied to a file when exporting/downloading it. There can be many use cases to apply a visible stamp like on classic paper. I get many PDFs from public health insurance companies with different stamps and we still use some stamps on paper because it is not yet possible to go completely paperless, but we have recreated the same stamps for Adobe Acrobat when we have a paperless workflow. We have to be able to make everything on paper and print it as it is displayed in the DMS or in the nursing software when an inspection authority such as the medical service of the health insurance companies, the nursing supervisory authorities, customs comes.
[automatically translated]

**Reply by madduck** [reply]
Yeah, definitely don't touch the original or archived document. Keeping these stamps and notes in the database means you can also generate a document that is a concatenation of the original along with a chronological paper trail of notes, and signatures. AdobeSign does this, you get a couple of pages worth of notes about when a document was sent to whom, when it was viewed, when it was signed, any notes, etc..

What is the state of upstream? What is their take on this?

**Comment by axgd-code** [comment]
Is this feature request related to https://github.com/paperless-ngx/paperless-ngx/discussions/4052 ?

DocuSign also provides an API for signing documents: https://github.com/docusign

**Reply by tinyoverflow** [reply]
No. This is not about cryptographical signing, but about the ability to create stamps, like you probably would on a physical piece of paper (like `PAID <DATE>`)

**Reply by axgd-code** [reply]
I think it's possible with DocuSeal (@AlexandrToorchyn)
![image](https://github.com/paperless-ngx/paperless-ngx/assets/1721781/e86c06be-20c1-48c2-9f96-f70448231698)
![image](https://github.com/paperless-ngx/paperless-ngx/assets/1721781/583f9520-e51a-4378-9a00-1ab14974f1b5)

**Reply by tinyoverflow** [reply]
DocuSeal is it's own solution and I don't think that an integration, if it ever happens, would be a proper solution to this feature suggestion. I think that most people, who simply want to stamp their documents, do not need a full blown PDF-form-creation and signing software.

**Comment by madduck** [comment]
No, my feature request is a lot more generic and specifically local and not at all related to some third party. I want to stamp a document, but the information should be cryptographically verifiable, using x509 or OpenPGP…

**Reply by axgd-code** [reply]
DocuSeal offers an alternative to DocuSign.
I initiated a discussion about it here: https://github.com/paperless-ngx/paperless-ngx/discussions/4052.
It would be great to consider developing a plugin to integrate the two solutions with API calls.

**Comment by calrissian13** [comment]
Custom stamps (e.g. "payed on xx") would be really great!

But as others already mentioned, it should not directly edit the stored PDF files. This should be done via database and only on the fly (like printing/downloading PDF as an option).

**Comment by tradexsrl** [comment]
maybe this could help: https://github.com/Stirling-Tools/Stirling-PDF/releases/tag/v0.20.1
has some api to integrate with other software

**Comment by CNordi** [comment]
An additional layer for stamps with dates, as mentioned above, short notes and markups would be greate.

**Comment by RoonniieeX** [comment]
I think this feature is really good! Please consider it in a future version.

**Comment by spencerandhill** [comment]
+1
Please bring this feature. A simple Set of customizable Stamps like "paid" or anything else the user could imagine would be great. You could do this with "pdf gear", but it's a local software and it doesn't work with paperless. So currently I have to export from paperless, do the stamp and re-import to paperless.

**Comment by MephistoJB** [comment]
I love this idea. It would really help me in my workflow

**Comment by ignoreigor** [comment]
why the hell would anyone want to do that? Mess up your documents with stamps? They may be markups that you can hide from the document, unlike smudging the originals with a pen, but where is the added benefit? The same task can be done better with tags, notes and user-defined fields. Why does a modern document archive have to reproduce every Stone Age workflow?

**Reply by MephistoJB** [reply]
Right now i do this by hand (written) on the document prior Scanning.. The benefit is That the Status of the document is visible in the document and therefore also outside of the DMS. One of the reasons I chose paperless was that it is able to keep documents in a file and folder structure so I can share this with other people who don't need to access the DMS.

**Reply by madduck** [reply]
> Right now i do this by hand (written) on the document prior Scanning.. The benefit is That the Status of the document is visible in the document and therefore also outside of the DMS. One of the reasons I chose paperless was that it is able to keep documents in a file and folder structure so I can share this with other people who don't need to access the DMS.

This means having to modify the document, which is not always acceptable, e.g. in Germany you are technically not allowed to annotate invoices.

You also cannot add other stamps later. And there is no history. And there is no identification of who made the stamp, nor any form of undeniability. It's a hack. I am advocating using cryptograpghic technology to do it properly.

**Reply by madduck** [reply]
> why the hell would anyone want to do that? Mess up your documents with stamps? They may be markups that you can hide from the document, unlike smudging the originals with a pen, but where is the added benefit? The same task can be done better with tags, notes and user-defined fields. Why does a modern document archive have to reproduce every Stone Age workflow?

please read https://github.com/paperless-ngx/paperless-ngx/discussions/3744#discussioncomment-6727145 and the rest of the discussion before getting outraged over single comments of people liking the idea.

**Reply by MephistoJB** [reply]
An income stamp for example is technically allowed as far as I know. And if it may not and you want to have audit proof storage. That's another business. And last but not least. Nobody needs to use it. But for those who would like this is a nice handy feature to save work

**Reply by madduck** [reply]
> An income stamp for example is technically allowed as far as I know. And if it may not and you want to have audit proof storage. That's another business. And last but not least. Nobody needs to use it. But for those who would like this is a nice handy feature to save work

It is not technically allowed. It is tolerated as long as e.g. placement and colour of ink make it clear that it's an addition. Anyway, this is a moot discussion. Cryptographic claims are so much more powerful and versatile, anything else I wouldn't advocate for, and especially nothing that actually modifies the original data.

**Reply by ignoreigor** [reply]
> please read [#3744 (comment)](https://github.com/paperless-ngx/paperless-ngx/discussions/3744#discussioncomment-6727145) and the rest of the discussion before getting outraged over single comments of people liking the idea.

There may be reasons to have this feature, that is requested. But one should have a look whether there are better functions to archive a goal - or if there are better tools to do something. 
Making paperless-ngx ending up in a full-featured pdf editor just to have visible markups that have some printable information binds resources that could be used for the further development of the main functionality or necessary reworks of the core.

**Comment by p001710** [comment]
Also ich wäre auch dafür diese Funktion zu implementieren. Ich habe aktuell eine Firma die sich so etwas wünscht und ich auf der Suche nach einer Möglichkeit.

**Comment by knebb** [comment]
I'd like to see such a feature as well.

I tried to create some scripted stamping mechanism.
My findings:

1. Tool to "stamp" pdf files:
Using [pdftk](https://manpages.debian.org/stretch/pdftk/pdftk.1.en.html) to "stamp" a pdf file. It consumes a *pdf file and merges it with a second stamp pdf file.
2. Create stamp file
I just created a simple pdf file through LibreOffice with a frame and 50% tranparency which will be merged with the to-be-stamped-pdf-file. This worked fine (see 3.) 
3. Stamp the target pdf with the given stamp file:
`pdftk input.pdf stamp stamp.pdf output result.pdf`

Some disadvantages:
* `pdftk`puts the stamp on EVERY page of the document instead of only the first page.
* The stamp is put just on top of the document, no matter if therre is some free space available for better readability.
* The stamp.pdf is not dynamic and has always the same content until manually changed.

To get a dynamical stamp I used `enscript`together with ps2pdf like this:
```
date >> text.txt
enscript --output=- text.txt | ps2pdf - stamp.pdf
```
This works fine, too. But the stamp.pdf is very simple and does not use any tranparency or frames. So the stamp will be less readable.

With these options you should be able to create a pre-/post-consumption script which will stamp your documents.

However, this does not suit me as I do not want all documents stamped. Only a subset of them. So I am looking for a possibility like the webhook to stamp them only when conditions apply. Any ideas about tthis?

I attached for your information the results.
[nicebutfixedstamp.pdf](https://github.com/user-attachments/files/19078231/nicebutfixedstamp.pdf)
[notsonicebutdynamic.pdf](https://github.com/user-attachments/files/19078232/notsonicebutdynamic.pdf)

**Comment by MephistoJB** [comment]
You can do this now with webhooks and Stirling PDF

https://app.swaggerhub.com/apis-docs/Frooodle/Stirling-PDF/0.43.2#/AddStampRequest

**Reply by knebb** [reply]
Ok, true.

But there is no way I am sending my confidential documents to an external public site.

If might be a good solution in case you need to stamp every document, indeed. But I guesss only for some documents this is a waste of ressources with a dedicated server application inside a docker...

**Reply by tradexsrl** [reply]
Stirling PDF is a self hosted docker service

**Reply by MephistoJB** [reply]
Exactly. The Website of linked to is only the api docu. You can keep it totally local. I would also not send my documents to an internet service

**Reply by knebb** [reply]
Sorry, I did not make it clear. I just mentionedd I am not going to send it to someone.

I was aware Sterling can be self-hosted but even then it might be a waste of ressources for (in my case) three or four documents per month....

**Reply by MephistoJB** [reply]
If you host paperless yourself Stirling can run on the same server. It needs practically no ressources.

Just as an idea how to solve this. I don't think that this will be realized within paperless. That objections against changing the document are very big in the community.

**Reply by axgd-code** [reply]
The fact of not modifying a document is an interesting point that stems from the topic of versioning. If Paperless-ngx ends up interfacing with other solutions, I believe this is an important point that will consistently come up. Moreover, if I modify a document, it might be useful to have the ability to revert back to a previous version.

**Reply by MephistoJB** [reply]
Paperless saves always the original file so I don't see a problem here

**Reply by R3dst4r** [reply]
Have you applied the Stirling solution in a real case scenario? Could you please describe your process how you solved it? 

My approach is to create a workflow which triggers a webhook that sends the PDF to stirling's endpoint (use curl to add-watermark or add-image (IIRC Stirling's add-image API endpoint is broken from day 1)). As a result, you can save the output file to the file system (e.g. export stirling=consume paperless) but how do you replace/update the original file (or the PDF/A) without having the version feature (yet)?

**Comment by OverDodo** [comment]
Since Workflows have grown so much recently, maybe it would be possible to add a Workflow Action similar to webhooks? Basically allow sending the pdf to an endpoint and expecting a pdf back, which would then be used as the archive file instead. Like a "Postprocessing" action.

That way we could freely use stirling and other Apis like signing and stamping. This would also allow removing passwords etc.


Pretty sure something like this was already requested, but I cant find it right now.

**Reply by tinyoverflow** [reply]
This sounds like an interesting idea. But can the workflows also modify the stored file? Even if this works, there might be the issue of stamp placement. As I imagine it, the workflow would have some default position or something. That might interfere with text on the document.

**Reply by madduck** [reply]
Please don't modify files. Cryptographic stamps can exist alongside the document without changing it. This way, multiple stamps can prove authenticity independently. If each physically altered the document, it would render all other signatures invalid.

If you need a stamp on the file later, take another tool that merges the document with the cryptographic data to create a new PDF outside of Paperless.

**Reply by OverDodo** [reply]
> This sounds like an interesting idea. But can the workflows also modify the stored file? Even if this works, there might be the issue of stamp placement. As I imagine it, the workflow would have some default position or something. That might interfere with text on the document.

Currently they cannot, but the archive file is already a modified version. And of course this should only modify the archive file. 
Placement etc. would need do be configured by the user, which is why I think a generic api call is much better than just the "stamp" feature.

**Reply by OverDodo** [reply]
> Please don't modify files. Cryptographic stamps can exist alongside the document without changing it. This way, multiple stamps can prove authenticity independently. If each physically altered the document, it would render all other signatures invalid.
> 
> If you need a stamp on the file later, take another tool that merges the document with the cryptographic data to create a new PDF outside of Paperless.

It should only change the archived version, which is already modified

**Reply by madduck** [reply]
https://github.com/paperless-ngx/paperless-ngx/discussions/3744#discussioncomment-6729586

**Reply by tinyoverflow** [reply]
> Please don't modify files. Cryptographic stamps can exist alongside the document without changing it.
This topic is not about cryptographic stamps. It is about physical stamps you can put on documents, like a physical "received at" or "paid at" stamp.

Regarding your statement about legal requirements for eg. Germany, yes. You're absolutely right. I'm also from germany. Though not everyone uses Paperless for business cases. I would assume most people use it for personal documents.

Doing such stamps would still be really nice, for example when required to print that document for whatever reason (happens sometiems). Then you have all the annotations on there. Maybe they could only be stored in the backend and put on the archived document when downloading (maybe with an explicit "download with annotations" options or something like this).

**Reply by madduck** [reply]
Exactly: combine the cryptographic claims with the document when the document is consumed, i.e. printed, forwarded, whatever. That way, you can have a blue stamp for the tax office and a green one of the accounting department. Or a different placement. Or location. Or not at all. Point is that you remain flexible until you know what you need, rather than taking those options away from you at a point when you don't need to take those options away.

**Comment by madduck** [comment]
Please do not modify files. There are strict requirements in parts of the world, PDF/1-a exists for a reason, and there is simply no need to merge the data at time of signing thanks to cryptographic techniques.

**Comment by p001710** [comment]
Hallo,
ich bin weiterhin sehr interessiert, habe aber keine Ahnung wie man das
umsetzen kann.

R3dst4r ***@***.***> schrieb am Di., 7. Apr. 2026, 15:28:
